### PR TITLE
Implement volume copy up.

### PR DESCRIPTION
**Pull request by ijc**
This pulls in and uses github.com/docker/docker/pkg/chrootarchive for the
actual copy up which is some battle hardened code to unpack avoiding things
like symlink traversal security issues.

However it does pull in a pretty huge pile of vendoring, including
github.com/docker/docker/pkg/reexec which we must then call at startup. It's
not immediately clear that this tradeoff is the correct one.

Signed-off-by: Ian Campbell <ijc@docker.com>

Relates to #186. With this change I can successfully login as one of the [predefined users](https://microservices-demo.github.io/docs/user-accounts.html) on the [sock-shop demo](https://github.com/microservices-demo/microservices-demo), previously this failed.

### CODE REVIEW

File: pkg/opts/container.go [hunk 1]
```diff
@@ -56,14 +57,36 @@ func WithVolumes(volumeMounts map[string]string) containerd.NewContainerOpts {
...
+		return err
+	}
+	if len(volList) > 0 {
+		srcList, err := ioutil.ReadDir(destination)
+		if err != nil {
+			return err
+		}
+		if len(srcList) == 0 {
+			// If the source volume is empty, copies files from the root into the volume
+			if err := chrootarchive.NewArchiver(nil).CopyWithTar(source, destination); err != nil {
```
Comment (Random-Liu) [inline pkg/opts/container.go]: Just for my education. Why do we need to chroot archive? I don't quite understand it.

File: pkg/opts/container.go [hunk 2]
```diff
@@ -56,14 +57,36 @@ func WithVolumes(volumeMounts map[string]string) containerd.NewContainerOpts {
...
 
+// copyExistingContents copies from the source to the destination and
+// ensures the ownership is appropriately set.
+func copyExistingContents(source, destination string) error {
+	volList, err := ioutil.ReadDir(source)
+	if err != nil {
+		return err
+	}
+	if len(volList) > 0 {
+		srcList, err := ioutil.ReadDir(destination)
```
[the remark this answers is not in the document]
Comment (ijc) [reply pkg/opts/container.go]: It confused me too but I stuck with it to keep in sync with Moby. Will fix.

File: pkg/opts/container.go (hunk 1 above)
Comment (ijc) [reply pkg/opts/container.go]: Mainly just:

> This pulls in and uses github.com/docker/docker/pkg/chrootarchive for the actual copy up which is some battle hardened code to unpack avoiding things like symlink traversal security issues.

This wouldn't be all that hard to implement directly but it does have some slightly tricky corner cases which have already been worked through in that code base. I think the `chroot` aspect of it is just to mitigate the security impact of any bugs.

It also covers a bunch of stuff I wouldn't have thought of myself, such as preserving xattrs and proper handling of hardlinks etc.

Comment (Random-Liu) [reply pkg/opts/container.go]: @ijc Good to know. Thanks for explaining this!

**Comment by Random-Liu** [comment]
LGTM

**Comment by Random-Liu** [comment]
We need an integration test for this behavior. Tracked this in #238.

**Pull request merged**